### ISSUE TITLE
Training pipeline update

### ISSUE THREAD

**Issue by Vandertic** [CuriosAI/sai#93]
Some powerful contributors recently started helping this project. While we thank and kudo them for their resources, some changes are needed for the training pipeline, as the AlphaZero pipeline requires some synchronization between training and self-plays.

Previously the self-plays were the bottleneck, and this allowed at each generation, some spare time for the training server. Hence we recently have followed a multiple-training approach: one main train of 5 networks, and then two or three additional trainings with 3 to 5 networks each, with modified hyper-parameters.

These multiple trainings allowed us to monitor the hyper-parameters and dynamically adjust them to get the better performances. In particular, the training window size and learning rate showed to be critical and we were convinced that they need to be checked and adjusted often.

When self-plays arrive so quickly as they are now arriving (but just for some generations), while the training server does all these trainings, the generation exceeds the required number of self-plays, that at the moment is 5120. This is somewhat unwanted, so we are doing some changes in the training pipeline.

1. When self-plays exceed a multiple of 5120, new self-plays are committed. Otherwise the subsequent games would all be branches and the proportion of whole games would decrease.
2. If self-plays exceed 5120, the trainings are limited to the main and the second one, hence for just 10 networks.
3. The number of training steps per generation will be dynamically adjusted to the increased number of available games. The rule of thumb is that the average training steps per generation should not exceed 1.25 times the average number of games per generation.

While I write, some of these changes are already automated, while others are hand-adjusted. Automation will be added in few days. If you see something that does not correspond to what written above for the first days, be patient.

**Comment by sheeryjay** [comment]
@Vandertic would it make sense to increase the number of visits to 3200? It would improve the quality/strength of the games and simultaneously somewhat lessen the impact of the resources.

Is it possible to contact the new contributors and perhaps ask them to wait few days? It seems that they use just few accounts (ionicliquids, maybe 13575860), yet the number of clients increased to 500 hundred from below 200. Does that mean that the clients are actually individual computes/IPs, perhaps ran by various individuals all reusing the same account/computer name (and making it impossible to actually contact them)?

I tried to look into the server.js to find out if it is possible to throttle/limit selfplayes when they reach say 10000, but on smartphone I could not find the proper place, if it is something you are interested in, I'll try to look in the weekend on when on PC.

**Comment by Vandertic** [comment]
We are really considering an increase in visits, but we want to wait a little bit more to see if the increased number of games will continue.
Also notice that it should really be enough to increase the training steps per generation to use this larger number of games.
Anyway, thank you for your offer.

**Comment by nanzi** [comment]
They are in Chinese go online chat group (QQ group) focusing on contributing.
Let me call them New-Powers. LeelaZ is anonymous, I haven't heard of them before neither.
 (OR because I don't focusing on contributing)

Today just earlier, I learnt that ionicliquids and 13575860 are both in the same chat group on SAI topic which I'm in too. SAI is much weaker than LeelaZ so far, only a few guys talks everyday in that group. Until recently some one tested SAI engine with lz257 on ladder reading and got SAI really famous.

Since GCP stopped his weights pipeline, some of New-Powers decided to quit and some of them decided to switch to other AIs like SAI or KataGo-distributive. That's all.

**Comment by Vandertic** [comment]
Awesome! Please, thank everybody on our behalf.

BTW, visits increased to 3500 from next generation. We plan one last increase to a (final?) value of 5000 after stepping to 12x256 and before the (final?) 20x256.

**Comment by l1t1** [comment]
http://www.yss-aya.com/cgos/19x19/cross/SAI_204_d8bf_p4000.html win a game against elfv2 p800

**Comment by yehud** [comment]
> Awesome! Please, thank everybody on our behalf.
> 
> BTW, visits increased to 3500 from next generation. We plan one last increase to a (final?) value of 5000 after stepping to 12x256 and before the (final?) 20x256.

hello,in Chain,we created a group online to training for leelazero, now,as you can see,Gcp didn't training networks for somedays and leelazero maybe stopped after a few days,so some of us join your project,such as ionicliquids,13575860,zhq,shengke,tkg1984,--import-this,yehud(me).....and some i can't remember now .in fact ,most of the training data is provide by us(may be 80%+),we will support your project before distributeing katago appears,after that,we will support the better one,so at least before distributing katago appears ,you can get  our greatest support.we usually using google colablike me,i registered 10 accounts,5 for lz,5 for sai(but the time i can use it was limited by google,only 2 hours or little a day) ionicliquids and 13575860 have hundreds of accounts ,come on!

**Comment by yehud** [comment]
> Awesome! Please, thank everybody on our behalf.
> 
> BTW, visits increased to 3500 from next generation. We plan one last increase to a (final?) value of 5000 after stepping to 12x256 and before the (final?) 20x256.

btw ,we tend to may be you should speed up training weights or take some measures now ,otherwise project efficiency will decrease

**Comment by Vandertic** [comment]
Yes, I am working on it

**Comment by l1t1** [comment]
pls update the line in homepage
Match games are played at full strength (only 3200 visits).
to 
Match games are played at full strength (3500 visits since net no.212?).

**Comment by kennyfs** [comment]
`Not each trained network will be a strength improvement over the prior one. Patience please. :)`
Also not applicable to sai.

**Comment by trinetra75** [comment]
@l1t1 Thanks. This is being fixed as I write.

@kennychenfs There is no guarantee that the promoted network is stronger than the previous. Even if the direct matches favour the newly promoted [note that not even this is a promotion requirement, even if it always has been so...] the effective difference in strength [as measured by the elo] will be set with the help of the reference matches and of all the other matches there of.

**Comment by kennyfs** [comment]
I mean the best net always changes in every generation, different from LZ sometimes the new net didn't pass.

**Comment by kennyfs** [comment]
hmm maybe it also ok for sai.

**Comment by Vandertic** [comment]
Hi all again,

we are finally done with rewriting all the scripts of the pipeline and they should be flexible enough to accommodate every number of clients and every possible speed in self-play.

Will rewrite the wiki page asap to give all the information.
